src/ZPublisher/WSGIPublisher.py

```diff
@@ -144,13 +144,22 @@ def _exc_view_created_response(exc, request, response):
             for key, value in exc.headers.items():
                 response.setHeader(key, value)
 
+        # Call the view so we can use it as the response body.
+        body = view()
+
         # Explicitly set the content type header if it's not there yet so
-        # the response doesn't get served with the text/plain default
-        if not response.getHeader('Content-Type'):
+        # the response does not get served with the text/plain default.
+        # But only do this when there is a body.
+        # An empty body may indicate a 304 NotModified from Plone,
+        # and setting a content type header will change the stored header
+        # in for example Varnish.
+        # See https://github.com/zopefoundation/Zope/pull/1088
```

Comment:
Refer to the issue and not the PR

```suggestion
        # See https://github.com/zopefoundation/Zope/issues/1089
```